tests/mock/vscode.ts

```diff
@@ -1024,31 +1093,73 @@ export class VSCode {
       await extension.activate();
 
     for (const [name, provider] of this._webviewProviders) {
-      const webview: any = {};
-      webview.asWebviewUri = uri => path.relative(this.context.extensionUri.fsPath, uri.fsPath);
-      const eventEmitter = new EventEmitter<any>();
-      let initializedPage: Page | undefined = undefined;
-      webview.onDidReceiveMessage = eventEmitter.event;
-      webview.cspSource = 'http://localhost/';
-      const pendingMessages: any[] = [];
-      webview.postMessage = (data: any) => {
-        if (!initializedPage) {
-          pendingMessages.push(data);
+      const { webview, pagePromise } = this._createWebviewAndPage();
+      provider?.resolveWebviewView({ webview, onDidChangeVisibility: () => disposable });
+      this.webViews.set(name, await pagePromise);
+    }
+  }
+
+  dispose() {
+    for (const d of this.context.subscriptions)
+      d.dispose();
+  }
+
+  async webViewsByPanelType(viewType: string) {
+    return (await Promise.all(this._webViewsByPanelType.get(viewType) ?? []));
+  }
+
+  async singleWebViewByPanelType(viewType: string): Promise<Page> {
+    const webViews = this._webViewsByPanelType.get(viewType);
+    if (!webViews?.size) {
+      while (true) {
```

Comment:
This makes it look quite messy. I would either await all webviews being created upon extension activation in the test mode, or would introduce a promise that resolves when all the webviews load, and then would expose the list of webviews w/ types or a getter by type.

Reply:
>  I would either await all webviews being created upon extension activation in the test mode

That function is not for `WebviewViewProvider` like `settingsView`, it's for `WebviewPanel` like the trace viewer embedded panel, and those are dynamic and asyncronous. One example of its usage, where one webview panel opens, closes, and a new webview panel opens again:

https://github.com/microsoft/playwright-vscode/blob/e782ce5e1fc9700419080bc847cd5099e6cb4fa7/tests/trace-viewer.spec.ts#L252-L268

Before I was using `expect.poll` on `webViewsByPanelType` before trying to get the webview because there's asynchronous calls that may delay the creation of the Trace Viewer panel (for instance, [vscode.env.asExternalUrl](https://code.visualstudio.com/api/references/vscode-api#env.asExternalUri)).

Reply:
I converted `singleWebViewByPanelType` into a test util function in 3505843, and there I simply perform a `expect.poll` to ensure there's a unique webview for a given panel type. I think it's a much simpler solution and achieves the same goal.